### ISSUE TITLE
Incorrect / confusing invalid-method-override when overriding str.__iter__

### ISSUE THREAD

**Issue by amartani** [astral-sh/ty#2612]
### Summary

The following code results in an `invalid-method-override` error on the `__iter__` method:

```py
from __future__ import annotations

from collections.abc import Iterator

class MyStr(str):
    def __iter__(self) -> Iterator[str]:
        raise NotImplementedError
```

https://play.ty.dev/a0d43982-dd9a-4f49-bdbf-8d75d867a595

mypy / pyright do not report an error; pyrefly does report a similar error.

This seems to happen due to the fact that `str.__iter__` is defined as an overload (with a `LiteralString`). Re-defining the overload clears the error:

```py
from __future__ import annotations
from typing import LiteralString, overload

from collections.abc import Iterator

class MyStr(str):
    @overload
    def __iter__(self: LiteralString) -> Iterator[LiteralString]: ...
    @overload
    def __iter__(self) -> Iterator[str]: ...
    
    def __iter__(self) -> Iterator[str]:
        raise NotImplementedError
```

But this seems something that shouldn't be necessary.

### Version

ty 0.0.13

**Comment by AlexWaygood** [comment]
Hmm, this is an interesting one -- thanks for the report!

The mypy comparison is not entirely a fair one, because mypy still has no support for `LiteralString` at all -- it patches typeshed so that these overloads on `str` just don't exist as far as mypy's concerned. You can see that happening in the fourth commit of their most recent typeshed sync, for example: https://github.com/python/mypy/pull/20605.

However, here's a minimal repro that doesn't require typeshed's stubs for `str`. You're correct that [ty](https://play.ty.dev/a3ec3ca2-e15b-403c-a1c6-f6ec29820d35) and [pyrefly](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeS4ATrgLYAEA%2Bg2AK4AuLlMTdENxuSmzoZ0uNqjYRc6OAB10YavTaliEdAHNe-QcNwA3GJSi5UmADR0Akm2OTBChQGMoqOHDoAxXLkQK6QLoAAUNjU3MAoMwYMDoaGDYAC1xMAAo4GCgwRDoAIVRKAEo6AFoAPhs7SgdKAG0CygBdXMI2qMDQoxMzTA66GLiE5NSMrJzvXxKKqvs2QTqfXBa6NsJ%2BwfjElPTM7OnK2zmFpaa6AB9Zmvn6xpb%2BoLoaiEy6ADlxa35YBPQ7TAAopRqJQnOhXO5PI1Wu10C43B58qgAF5pJZFfzoR6bYY7Mb7MqHaq1Ra%2Be5Yx5BZ6vD5sL7EH4wP4wQHAxzoEAWEDsaBwEjkRAgADEdAAqlIoBBVHRWOCpDJ5HCsLEZYIaJIGOgWDRsMY0vhcho2Ac6HA2JRMY8uBxKFiwHIQG9tbqLXRgPgAL4OhSckBkLhgKCkQjzGhQCgigAKpADQdNGBwBDozhkkE0nEk0nQ63QIoAyjAYHQkmw2MQ4IgAPSV-2xIOEQSaStMyuYXDOOCVlOKCDp65ZyuqygiAyoaCobCwZOp3sZ%2BVY3DEed8hRkEboUrdOBZugAXjoDoAzIQAIwAJm9HI9XNQzikRi80BgFDQWDwRDIIA9QA) both issue a Liskov violation error on `Baz.method` here, while [mypy](https://mypy-play.net/?mypy=latest&python=3.12&gist=9c7b83d3104aa414bd2a9eeb6c3e2974) and [pyright](https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.14&pythonPlatform=Darwin&reportUnreachable=true&code=GYJw9gtgBA%2BjwFcAuCQFM5QJYQA5hCSgEMA7UsJYpLMUgZwChRIokBPXLUgc2zwJEwANzQgANmGIATADRQAkkjHUCjRgGNxxevSgAxMGABcjKOagABEWMkyzF6WmBQIaJAAsw0gBT004sDGUABCxCAAlFAAtAB8isogqiAA2mEgALrBAHS5DubWohJS0vlQTi5unt5%2BAUEGRlFxCSpIBCmGYFlQudllFa7uXr7%2BgU3xSq3tnRlQAD4tSW2p6VllFlBJWP5QAHKUCnjiaG6kytIAoiDgIOpaOnrpOXma2rqhxABePp0RphsDKrDWpjGITRLJDpGNYbDZbHb7JCHXDHU7nK43RhAA) do not:

```py
from __future__ import annotations
from typing import overload, Iterator

class Foo:
    @overload
    def method(self: Bar) -> Iterator[Bar]: ...
    @overload
    def method(self: Foo) -> Iterator[Foo]: ...
    def method(self) -> Iterator[Foo] | Iterator[Bar]:
        raise NotImplementedError

class Bar: ...

class Baz(Foo):
    def method(self) -> Iterator[Foo]:
        raise NotImplementedError
```

**Comment by carljm** [comment]
I think the simplified repro and the original case are somewhat different, in that a common subclass of `Baz` and `Bar` is possible (which makes the error arguably correct in that case), whereas a common subtype of `MyStr` and `LiteralString` is not possible, making the error clearly wrong in the original case; the `LiteralString` overload can be ignored entirely by `MyStr` because the `MyStr` override can never be called with a `self` assignable to `LiteralString`.

These self-distinguished overloads are pretty weird, because the more sensible thing would be to override `method` on `Bar`, not overload it on `Foo` based on the type of `self`. But typeshed can't do this because there is no place to define methods on `LiteralString`. Arguably typeshed would be cleaner if it had a `LiteralStringFallback` class definition which inherits `str` and overrides some methods, rather than these overloads.

**Comment by carljm** [comment]
The fix here, I think, would be to ignore base class overloads whose `self` type is disjoint from the current class, when doing method override checks.

**Comment by carljm** [comment]
Closing this as dupe of #2693, which has more discussion and examples; they are discussing the same issue.
